internal/characters/layla/shield.go

```diff
@@ -0,0 +1,209 @@
+	n.Tmpl.Expires = c.Core.F + dur
+	n.c = c
+	return n
+}
+
+func (c *char) AddNightStars(count int, icd ICDNightStar) {
+	if c.Tag(shootingStars) > 0 {
+		return
+	}
+
+	switch icd {
+	case ICDNightStarSkill:
+		if c.StatusIsActive(starSkillIcd) {
+			return
+		}
+		c.AddStatus(starSkillIcd, 0.3*60, false)
+	case ICDNightStarBurst:
+		if c.StatusIsActive(starBurstIcd) {
+			return
+		}
+		c.AddStatus(starBurstIcd, 0.5*60, false)
+	}
+
+	if c.a1Stack < 4 {
+		c.a1Stack++
+	}
+
+	stars := count + c.Tag(nightStars)
+	if stars > 4 {
+		stars = 4
+	}
+	c.SetTag(nightStars, stars)
+	c.Core.Log.NewEvent("adding stars", glog.LogCharacterEvent, c.Index).
+		Write("stars", stars)
+
+	if stars == 4 && c.Tag(shootingStars) == 0 {
+		c.SetTag(shootingStars, 1)
+		c.shootStarSrc = c.Core.F
+		c.Core.Tasks.Add(c.ShootStars(c.shootStarSrc, -1), 0.5*60)
+	}
+}
+
+func (c *char) ShootStars(src int, last int) func() {
+	return func() {
+		if c.shootStarSrc != src {
+			return
+		}
+		if c.Tag(shootingStars) == 0 {
+			return
+		}
+
+		// find near target
+		nearTarget := -1
+		x, y := c.Core.Combat.Player().Pos()
+		trgs := c.Core.Combat.EnemiesWithinRadius(x, y, 10)
+		if len(trgs) > 0 {
+			sort.Slice(trgs, func(i, j int) bool { return i < j })
+			nearTarget = trgs[0]
+		}
+
+		if last == -1 {
+			// if not found
+			if nearTarget == -1 {
+				c.Core.Tasks.Add(c.ShootStars(src, -1), 0.1*60)
+				return
+			} else {
+				c.Core.Tasks.Add(c.ShootStars(src, nearTarget), 0.5*60)
+				return
+			}
+		}
+		if nearTarget == -1 {
+			nearTarget = last
+		}
+
+		stars := c.Tag(nightStars)
+		if stars == 4 && c.Base.Cons >= 4 {
+			for _, char := range c.Core.Player.Chars() {
+				char.AddStatus(c4Key, 3*60, true)
+			}
+		}
+
+		ai := combat.AttackInfo{
+			ActorIndex: c.Index,
+			Abil:       "Shooting Star",
+			AttackTag:  combat.AttackTagElementalArt,
+			ICDTag:     combat.ICDTagElementalArt,
+			ICDGroup:   combat.ICDGroupLayla,
+			StrikeType: combat.StrikeTypeDefault,
+			Element:    attributes.Cryo,
+			Durability: 25,
+			Mult:       starDmg[c.TalentLvlSkill()],
+			FlatDmg:    0.015 * c.MaxHP(), // A4
+		}
+
+		done := false
+		cb := func(_ combat.AttackCB) {
+			if done {
+				return
+			}
+			done = true
+
+			if !c.StatusIsActive(skillEnergy) {
+				var count float64 = 1
+				if c.Core.Rand.Float64() < 0.33 {
+					count = 2
+				}
+				c.Core.QueueParticle("layla", count, attributes.Cryo, c.ParticleDelay)
+				c.AddStatus(skillEnergy, 3.5*60, true)
+			}
+			if c.Base.Cons >= 2 {
+				c.AddEnergy("layla-c2", 1)
+			}
+		}
+
+		c.Core.QueueAttack(ai, combat.NewCircleHit(c.Core.Combat.Enemy(nearTarget), 0.5), 0, starsTravel[len(starsTravel)-stars], cb)
+
+		stars--
+		c.SetTag(nightStars, stars)
+		if stars == 0 {
+			c.RemoveTag(shootingStars)
+			c.starTickSrc = c.Core.F
+			c.TickNightStar(c.starTickSrc, false)
+		} else {
+			c.Core.Tasks.Add(c.ShootStars(src, nearTarget), 0.45*60)
+		}
+	}
+}
+
+func (c *char) TickNightStar(src int, star bool) func() {
```

Comment:
Capital letter in the beginning changes visibility, so please rename it to tickNightStar.